saleor/payment/gateways/braintree/__init__.py

```diff
@@ -136,128 +115,109 @@ def get_client_token(**connection_params: Dict) -> str:
     return client_token
 
 
-def authorize(
-        payment: Payment,
-        payment_token: str,
-        **connection_params: Dict) -> Tuple[Transaction, str]:
+def process_payment(
+        payment_information: Dict, **connection_params: Dict) -> Dict:
+    auth_resp = authorize(payment_information, **connection_params)
+    payment_information['token'] = auth_resp['transaction_id']
+
+    try:
+        assert auth_resp['is_success']
+        return [auth_resp, capture(payment_information, **connection_params)]
+    except Exception:
+        return [auth_resp, void(payment_information, **connection_params)]
+
+
+def authorize(payment_information: Dict, **connection_params: Dict) -> Dict:
     gateway = get_braintree_gateway(**connection_params)
+
     try:
         result = gateway.transaction.sale({
-            'amount': str(payment.total),
-            'payment_method_nonce': payment_token,
+            'amount': str(payment_information['amount']),
+            'payment_method_nonce': payment_information['token'],
             'options': {
                 'submit_for_settlement': CONFIRM_MANUALLY,
                 'three_d_secure': {
                     'required': THREE_D_SECURE_REQUIRED}},
-            **get_customer_data(payment)})
+            **get_customer_data(payment_information)})
     except braintree_sdk.exceptions.NotFoundError:
-        return transaction_and_incorrect_token_error(
-            payment, kind=TransactionKind.AUTH, token=payment_token)
+        raise BraintreeException
+
     gateway_response = extract_gateway_response(result)
     error = get_error_for_client(gateway_response['errors'])
-    credit_card_data = gateway_response.pop('credit_cart', None)
-    if credit_card_data:
-        payment.cc_first_digits = credit_card_data.bin
-        payment.cc_last_digits = credit_card_data.last_4
-        payment.cc_brand = credit_card_data.card_type
-        payment.cc_exp_month = credit_card_data.expiration_month
-        payment.cc_exp_year = credit_card_data.expiration_year
-        payment.save(update_fields=[
-            'cc_first_digits', 'cc_last_digits', 'cc_brand',
-            'cc_exp_month', 'cc_exp_year'])
-
-    txn = create_transaction(
-        payment=payment,
-        kind=TransactionKind.AUTH,
-        amount=gateway_response.pop('amount', payment.total),
-        currency=gateway_response.pop('currency', payment.currency),
-        gateway_response=gateway_response,
-        token=getattr(result.transaction, 'id', ''),
-        is_success=result.is_success)
-    return txn, error
-
-
-def capture(
-        payment: Payment,
-        amount: Decimal,
-        **connection_params: Dict) -> Tuple[Transaction, str]:
+
+    return {
+        'is_success': result.is_success,
+        'kind': 'auth',
+        'amount': gateway_response.get('amount'),
```

Comment:
It worth to mention that the gateway should only try to provide the ```amount``` and ```currency``` if they are truely in the response. They shouldn't provide default values such as ```payment.total```. The saleor's side will take care of it by providing the default values such as by ```gateway_response.get('amount', payment_information['amount'])```

Reply:
If we decide the default values of amout and currency should be provided from the gateway side, then we should take them as required fields in ```validate_gateway_response``` and do not provide default values in saleor's ```create_transaction```

Reply:
My thinking was that the gateway always provides values for required fields. Be it from the response of the provider api or reused values from what the gateway received if there's none from api response. I added extra validation to `validate_gateway_response` with type checking as well. I also made the response consistent in all gateways.